kythe/docs/schema/modules.txt

```diff
@@ -0,0 +1,148 @@
+
+[source,javascript]
+----
+//- @"'./module'" ref/imports Module
+import './module';
+//- @module ref Module
+module.foo();
+----
+
+With these edges, users who attempt to navigate from the module's use site will
+be brought to the definition of `Module`, not to the definition of the local
+alias. In the same way, users investigating cross-references from the module's
+definition will see every use of that module in the database (up to re-exporting
+and vexing control-flow), not just the places where it's been imported. In
+addition, other information (like documentation) will be available at every use
+site.
+
+Some languages allow particular values to be imported into local scope. These
+should be handled in the same way as module definitions:
+
+[source,javascript]
+----
+//- @value ref/imports Value
+import {value} from './module';
+//- @value ref Value
+value;
+----
+
+== Imports with renaming
+
+It's sometimes possible to choose a different name for the local binding created
+by an import from the target definition's. In this case, the local name *does*
+have a sensible syntactic location and is semantically relevant to the
+programmer. A use site of the local name refers to both the module (or value)
+being imported and to the local alias:
+
+[source,javascript]
+----
+//- "./module'" ref/imports Module
+//- @mod_imp defines/binding ModImp
+//- ModImp.subkind import
+import * as mod_imp from './module';
+//- @value ref/imports Value
+//- @renamedValue defines/binding RenamedValue
+//- RenamedValue.subkind import
+import {value as renamedValue} from './module';
+//- @mod_imp ref Module
+//- @mod_imp ref ModImp
+mod_imp.foo;
+//- @renamedValue ref Value
+//- @renamedValue ref RenamedValue
+renamedValue;
+----
+
+== Re-exporting and rebinding
+
+If a module re-exports a definition (renaming it or not) that it previously
+imported, that re-export site should be considered to be the definition of that
+name.
+
+If a program rebinds an imported name, indexers should make a best effort to
+*ref* the corresponding module:
+
+[source,javascript]
+----
+//- @"'./module'" ref/imports Module
+import './module';
+//- @module ref Module
+module.foo();
+//- @"'./debug/module'" ref/imports DebugModule
+import './debug/module';
+//- @module ref DebugModule
+module.foo();
+----
+
+This is not generally possible to do. In cases where the corresponding module
+can't be determined, indexers should fall back on treating imports as locals.
+This may require some abuse of syntax to find definition sites where none exist:
+
+[source,javascript]
+----
+//- @"'./module'" ref/imports Module
+//- @import defines/binding ModuleImport
+//- ModuleImport.subkind import
+import './module';
+//- @module ref ModuleImport
+if (random()) module = otherModule;
+//- @module ref ModuleImport
+module.foo();
+----
+
+== Caveats
+
+Downstream consumers of the graph should tolerate overlapping references.
+Variables defined by imports should have subkind `import`. Consumers may
```

Comment:
Should this be the `subkind` fact?  Wouldn't this be awkward for imports of, for example, classes since they already have a subkind?

Reply:
This treatment uses `variable` nodes with `import` subkind to represent the local binding to the thing that was imported (e.g., a class).

Reply:
That seems somewhat reasonable.  I might prefer making `import` it's own `kind`.

At the very least we'd need a suggested node for when the indexer doesn't have a better kind to give the import?

Reply:
I think that a `variable` with the `import` subkind will work in the general case. I don't think it's different enough to warrant a new top-level kind.

Reply:
In languages with precise information on the kind of an imported name and more specific node kind (such as `talias`) do we want to recommend they use that and only fall back to `variable+import`?

Reply:
variable+import is precisely a name introduced as part of an import (that may point to something of a different kind).